### ISSUE TITLE
Initial impressions

### ISSUE THREAD

**Issue by 0323pin** [jzbor/marswm#1]
Hi,

Came across this project while browsing crates.io yesterday. As a **LeftWM** user for over a year now, I was curious to test this one :)

_Side note: I'm a pkgsrc/NetBSD maintainer, so I've made a package for `marswm` and here it is running on my machine_,

![2023-03-23-092435_1366x768_scrot](https://user-images.githubusercontent.com/90570748/227147333-b8188265-592c-4acb-9ae2-938c417e9fd7.png)

I know you didn't ask for it but, here are my initial thoughts.

Generally, the WM works well. Related to `marsbar`, I'm missing the documentation for `marsbar.yaml` which, I would like to configure somewhat.
Yes, I can read `marswm/marsbar/src/config.rs` and work my way through it but, it would be nicer with an example config.

I'm planning to use `marswm` for a week to test it. Let me know if you would like me to wait with merging the package into our repositories or, if you are fine with me doing that after testing.

Can get back to you if I find any issues.

**EDIT:** `marsbar print-default-config` works :)

**Comment by jzbor** [comment]
Hi,
Thanks for the feedback! It is great so see someone actually using it other than myself, so also thanks for sharing the screenshot. 

I know that the documentation for `marsbar` is sadly lacking at the moment. Also (in contrast to `marswm` I hope) there are still some quite substantial bugs to be fixed. For example the small dark square on the top left corner is actually the status widget that is not quite properly initialized when there is no status. The red square is an empty system tray. I was not aware however of the tearing issue you mentioned initially.

If you are interested in how a status script looks you can read/test/extend mine, which is currently available [here](https://github.com/jzbor/home-config/blob/master/scripts/mars-status.sh). I will improve `marsbar` documentation with the next bigger release.

Speaking of the next release I plan on improving the argument parsing on all the programs, which might change the commandline interface a little, but the functionality should remain the same. I currently have exams at university, so I will probably have time for working on this in about two weeks.

Thanks again for your feedback - it is much appreciated :)

**Comment by jzbor** [comment]
> I've just pushed the package to our WIP repository, http://mail-index.netbsd.org/pkgsrc-wip-changes/2023/03/23/msg026677.html But, I'll wait for your improvements before merging it into the main branch.

Yes I think that makes sense
>
> If it's not asking too much, it would be nice if the bar height could be adjusted, I didn't see an option for it and, as you can see from the screenshot in unnecessarily _fat_ :)

Ouh yeah that it definitely a missing feature, that needs to be implemented.

> As comparison, `lemonbar` on LeftWM bellow ![2023-03-23-131624_1366x768_scrot](https://user-images.githubusercontent.com/90570748/227201647-6c20ff6f-76a0-4bfb-b868-aaac3bde76a9.png)

You _should_ be able to use lemonbar with marswm, although I haven't tested it and these bars are always a little tricky.

> Good luck with the exams!

Thanks :)

**Comment by 0323pin** [comment]
> You should be able to use lemonbar with marswm, although I haven't tested it and these bars are always a little tricky.

Yeah but, I would prefer to use `marsbar`. `lemonbar` works ok on LeftWM after some filtering through `liquid`, in case you'd be [interested](https://codeberg.org/pin/leftwm-netbsd/src/branch/main/.config/leftwm/themes/poppy)

**Comment by 0323pin** [comment]
Awesome, thanks!

>  I have also added an option to configure the bar height.

Nice. For now, I'm running without bar.

![2023-03-25-081225_1366x768_scrot](https://user-images.githubusercontent.com/90570748/227702979-035bec79-c78c-4206-9636-177f82e35841.png)

The bar height it working but, I think adding the possibility to set the **font** and **font size** would make it more attractive.

Nice progress. I'm enjoying this WM.

**Comment by jzbor** [comment]
> The bar height it working but, I think adding the possibility to set the **font** and **font size** would make it more attractive.

Yeah I know, I thought of this yesterday. I will definitely add it. Although the way it it is implemented right now (Xft) I found it to be quite picky on the font/size combos it accepts, so that might be an issue even when the config option is already in place.

**Comment by 0323pin** [comment]
> Although the way it it is implemented right now (Xft) I found it to be quite picky

Yeah, I know. I'm using [lemonbar-xft](https://gitlab.com/protesilaos/lemonbar-xft/)

**Comment by 0323pin** [comment]
@jzbor Everything seems fine, been using it all day.

**Comment by jzbor** [comment]
> @jzbor Everything seems fine, been using it all day.

Nice, I am glad to hear that.

I have also added a font option for marsbar now, so you are welcome to try it out.

**Comment by 0323pin** [comment]
> I have also added a font option for marsbar now, so you are welcome to try it out.

Is it also possible to define the font size?

I've re-built `marswm` from git-head again but, still need to try `marsbar`, been playing around with something else :)

Here's `marswm` as the widow-manager  for LXQt,
![2023-03-26-104602_1366x768_scrot](https://user-images.githubusercontent.com/90570748/227768675-bae1df14-dfba-4740-8cc0-d6ec2bcf53a8.png)

**Comment by jzbor** [comment]
> Is it also possible to define the font size?

Yes you can specify it along with the font name (although it sometimes doesn't work - I'll blame it on xft):
```yaml
  font: FiraCode:size=12           # font of text surfaces (as xft name)
```

> Here's marswm as the widow-manager for LXQt,

Ouh that looks nice - does it integrate well with the rest of LXQt?

**Comment by 0323pin** [comment]
> does it integrate well with the rest of LXQt?

Let's say it's usable. Initially, everything seemed to work fine, as long as the `lxqt-panel` was at the _bottom_ or _top_ of the desktop and `lxqt-globalkeys` was disabled. 

Placing the panel to the _left_ resulted in the whole desktop freezing and not responding to any keybindings.

There are a few glitches here and there, like _Mod4+Return_ doesn't launch a terminal but, it does start a second one if I launch the first one from the menu. But, in general works.

I always feel like I have a _dirty_ system whenever I'm using a full DE, so I've removed it now. Perhaps, I should try this on Linux on a VM.

**Comment by 0323pin** [comment]
Working just fine :)

![2023-03-26-182800_1366x768_scrot](https://user-images.githubusercontent.com/90570748/227789940-3ad215a8-c6b2-41db-8ef2-20838c6eab08.png)

**Comment by jzbor** [comment]
> Placing the panel to the _left_ resulted in the whole desktop freezing and not responding to any keybindings.

Yes marswm only handles bottom or top panels for now...

> There are a few glitches here and there, like _Mod4+Return_ doesn't launch a terminal but, it does start a second one if I launch the first one from the menu. But, in general it works.

Hugh thats weird, can think of anything right now that would cause this...

**Comment by 0323pin** [comment]
> thats weird, can think of anything right now that would cause this

I would say, this needs to be tested on the full DE, I only built a selected set of components and swapped-out `openbox`.

If I find the time, I will try this on Linux, there's a old brick in the house that has Void with LXQt but, it hasn't been used in months.

**Comment by 0323pin** [comment]
@jzbor 

> there's a old brick in the house that has Void with LXQt

So, I couldn't resist and had to take it for a spin :)

![2023-03-27-064137_1366x768_scrot](https://user-images.githubusercontent.com/90570748/227864861-1b880fbc-d702-4cdf-bde8-de9df9f3571e.png)

It works pretty well but, the weird thing with the keybindings remains. I did switch off `lxqt-globalkeys` in order to avoid potential conflicts with already defined keys but, I get the same behavior. Keybindings are only functional on workspaces that already have one client open.

For example, if I launch an application from the menu (anyone would do), then I can use keybindings to open others but, using any keybinding on a workspace without any client does nothing. Actually, if a workspace is empty it's not even possible to switch to another workspace, on the other hand, if there's an open application this works.

**Comment by jzbor** [comment]
> as desktop control is still in the hands of `pcmanfm-qt`.

I think this causes the problem with the hotkeys: pcmanfm sets a desktop window, which `marswm` simply maps and ignores. But this means that on an empty workspace there is still that pcmanfm window above the root window, blocking the root window from getting key events...

Better handling for unmanaged windows is still an issue in general, another example is windows getting raised above notifications. But it is good to know this...

**Comment by jzbor** [comment]
> ```
> 4
> ```
> 
> 
>     
>   
> 
> The issue is with the `status_cmd`. If I set `status_cmd: mars-relay set-status "$(date +%F)"` and load the bar it shows the date. I can then change this to `status_cmd: mars-relay set-status "$(date +'w%W %a %F')"` and reload and it still works but, if I start with the later option, it displays just an empty square.

Hm so I think if you get only one square that is probably just the system tray and there is no status found at all. The problem with your command is that it only sets the status once. The `status_cmd` is only executed once when the bar is launched.

There is a bug in marsbar that I have not been able to pin down yet, where it does not always pick up the status when it is set. That is usually no problem as the next time the script updates the status it picks it up correctly. In your case though, where the status is only set once it somehow misses it and then never gets a notification about that property ever again.

You can check what is set as status by checking the `_MARS_STATUS` property on the root window.

**Comment by 0323pin** [comment]
> Hm so I think if you get only one square that is probably just the system tray

Yes, just one. How is the system tray supposed to work? Or, better, I'll wait for the documentation.

> The problem with your command is that it only sets the status once

True but, I thought that would be enough. I have the time on my shell prompt and the date doesn't change so quickly :)
I'll look into creating a loop.

**Comment by jzbor** [comment]
> How is the system tray supposed to work?

It should just pick up tray applications you are using, like nm-applet, nextcloud, whatever. If you don't have any there is nothing to show, although I'll admit it would be nice to hide the tray in that case...

> True but, I thought that would be enough.

Yeah it should be - it is definitely a bug (I suspect racing conditions).

**Comment by 0323pin** [comment]
> It should just pick up tray applications you are using, like nm-applet, nextcloud, whatever. If you don't have any there is nothing to show

Everything is behaving as it should then. I don't use this sort of things :)

**Comment by 0323pin** [comment]
> It should just pick up tray applications you are using, like nm-applet, nextcloud, whatever. If you don't have any there is nothing to show

Everything is behaving as it should then. I don't use this sort of things :)

**Comment by 0323pin** [comment]
One last thing for now. The following is `lemonbar` with `leftwm`

![workspaces](https://user-images.githubusercontent.com/90570748/227943532-7de7f2bf-bd7f-4892-bb9f-aecac0d23dce.png)

The **green** label on the workspace means it's the currently active one. The **yellow** label indicate that there are clients open on those workspaces. Is this something you have plans to add as a feature?

I think after this, I don't think I'd need any further features :)

**Comment by jzbor** [comment]
> The **green** label on the workspace means it's the currently active one. The **yellow** label indicates that clients are open on those workspaces. Is this something you have plans to add as a feature?

Yeah I had something like this with my old dwm setup. I would definitely like to implement it, but it requires tracking the windows from the `marsbar`, so it is on the backlog for now. I would really like this, so I would probably implement it eventually.

As for some of the other issues: I have fixed the status not getting set the first time, so your `date` command from above should work now. Also the system tray is hidden, when there are no tray applications.

**Comment by 0323pin** [comment]
Been using the latest additions for sometime now. Things are smooth, thanks for the fixes.

**Comment by 0323pin** [comment]
Would you happen to know where and how should I define media keys?
On LeftWM, I have the following,

```
        (command: Execute, value: "mixerctl -w outputs.master-=12", modifier: [], key: "XF86XK_AudioLowerVolume"),
        (command: Execute, value: "mixerctl -w outputs.master+=12", modifier: [], key: "XF86XK_AudioRaiseVolume"),
```

But, adding something similar to keybindings doesn't work. Should this be added to the buttonbindings instead?

**Comment by jzbor** [comment]
This is what I am using currently:
```yaml
# Media Control
- key: XF86AudioPlay
  action: !execute playerctl play-pause -p spotify
- key: XF86AudioPause
  action: !execute playerctl play-pause -p spotify
- key: XF86AudioPrev
  action: !execute playerctl previous -p spotify
- key: XF86AudioNext
  action: !execute playerctl next -p spotify
- modifiers: [ Mod4 ]
  key: i
  action: !execute playerctl play-pause -p spotify
- modifiers: [ Mod4 ]
  key: u
  action: !execute playerctl previous -p spotify
- modifiers: [ Mod4 ]
  key: o
  action: !execute playerctl next -p spotify
```

**Comment by 0323pin** [comment]
Thanks!

> You can find out the name of a specific key with xev

Yeah, this I'm aware of but, I was trying:

```
modifiers: []
- key: XF86AudioRaiseVolume
  action: !execute mixerctl -w outputs.master+=12
```

**Comment by jzbor** [comment]
Ah I think I see whats wrong: The modifiers must be part of a specific key binding. So something like this should work:
```yaml
- key: XF86AudioRaiseVolume
  modifiers: []
  action: !execute mixerctl -w outputs.master+=12
```
If I am not mistaken removing the `modifiers` field completely should also work:
```yaml
- key: XF86AudioRaiseVolume
  action: !execute mixerctl -w outputs.master+=12
```

**Comment by 0323pin** [comment]
[WIP](https://codeberg.org/pin/marswm) 😄

**Comment by 0323pin** [comment]
[Got the volume keys working](https://codeberg.org/pin/marswm/commit/30b02be6985f38a726871574955e439aaa2cf81b). None of the other options did it.

**Comment by 0323pin** [comment]
@jzbor https://www.reddit.com/r/unixporn/comments/12cikkt/marswm_another_planet/

**Comment by jzbor** [comment]
Looks great! I didn't even think about a titlebar-ish border on the side. It's nice to see it enable configurations that I didn't even plan for...

**Comment by 0323pin** [comment]
Closing this now. let's open a new issue when needed.

**Comment by 0323pin** [comment]
> I have now published a new version as 0.4.0

🎉 

> There are still some issues, but no deal breakers so I think it should be fine for packaging in pkgsrc

True, I've been using git-HEAD for the past week and it's ready for prime time. I'll move the package to the main branch today. I'll post here when it's done and probably create a PR with install instructions.

**Comment by 0323pin** [comment]
@jzbor Merged now, http://mail-index.netbsd.org/pkgsrc-changes/2023/04/12/msg272363.html
